### PR TITLE
Adding ability to output PF Candidate information in JME NTuples

### PR DESCRIPTION
**Pull request by laurenhay**
#### PR description:

Adding ability to output PF Candidate information belonging to a desired jet collection along with a mapping between said jets and candidates to the custom JME workflows.

Related materials:
NanoAODJMAR github: [https://github.com/cms-jet/NanoAODJMAR](https://github.com/cms-jet/NanoAODJMAR)
JME Twiki: [https://twiki.cern.ch/twiki/bin/view/CMS/JMECustomNanoAOD](https://twiki.cern.ch/twiki/bin/view/CMS/JMECustomNanoAOD)
PR for NanoAODJMAR's candidate tables: #29550
Last JMAR meeting slides announcing JME+PF: [https://indico.cern.ch/event/962252/contributions/4047997/attachments/2116746/3562198/JMEBTVnanoDiscussion_20201006_updated.pdf](https://indico.cern.ch/event/962252/contributions/4047997/attachments/2116746/3562198/JMEBTVnanoDiscussion_20201006_updated.pdf)

#### PR validation:

Added runTheMatrix workflows:
`25202.16`
which adds only AK8 PF Candidates to the JME workflow `25202.15`, and:
`25202.17`
which adds all PF Candidates to `25202.15`.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

Tagging @rappoccio

### CODE REVIEW

**Comment by mariadalfonso** [comment]
-1
There is also no consensus on how to flatted the edm objects
we have not discussed how technically proceed with the multiplication of flat tree for Run3 developments
 
the jmeNano is currently unattended.
when experts updated the central nano, lead to crash in the JMEtree.
https://github.com/cms-sw/cmssw/pull/31446 (from Lauren JME)
https://github.com/cms-sw/cmssw/pull/31786 (from BTV)


The nanoGen is much more structured and should inspire additional tree for PPD and POGs

**Comment by laurenhay** [comment]
For more context, this is what some sample contents are:
https://lahay.web.cern.ch/lahay/website2/PFJME_JetHT2017UL_size.html
https://lahay.web.cern.ch/lahay/website2/PFJME_JetHT2017UL_description.html
https://lahay.web.cern.ch/lahay/website2/AK8PFJME_JetHT2017UL_size.html
https://lahay.web.cern.ch/lahay/website2/AK8PFJME_JetHT2017UL_description.html
https://lahay.web.cern.ch/lahay/website2/JME_JetHT2017UL_size.html
https://lahay.web.cern.ch/lahay/website2/JME_JetHT2017UL_description.html
html files starting with PFJME are using the custom JME and have all PF cands stored, AK8PFJME is only the fatjet candidates, and JME is with no PF cands.
The configuration files for these samples were created using the following command in 11_2_pre7:
```
cmsDriver.py step1 --data --eventcontent NANOAOD --datatier NANOAOD --conditions auto:run2_data --step NANO --era Run2_2017,run2_nanoAOD_106Xv1 --python_filename PFJME_Data17UL_cfg.py --no_exec --customise_commands="process.add_(cms.Service('InitRootHandlers', EnableIMT = cms.untracked.bool(False))) \n from PhysicsTools.NanoAOD.custom_jme_cff import PrepJMECustomPFNanoAOD_Data; PrepJMECustomPFNanoAOD_Data(process)\n "
```
And then directly editing the config file to change the input file to:
```
/store/data/Run2017B/JetHT/MINIAOD/09Aug2019_UL2017-v1/130000/07017B4E-9794-DB40-BFBE-F8664579C066.root
```
and the number of events to be 10000.
This will not exclusively be used for a central production, and is also being used by BTV and JMAR for private samples.

**Comment by chayanit** [comment]
Hello @laurenhay, why do we need new workflows for all 3 MC year? Is it supposed to be included in IB test only?

**Comment by rappoccio** [comment]
> Hello @laurenhay, why do we need new workflows for all 3 MC year? Is it supposed to be included in IB test only?

Hi @chayanit, do you mean [here](https://github.com/cms-sw/cmssw/pull/31795/files#diff-8c163203cbaa64f0e0e6aa2b16346cfab3eaf84a57a008c7fa37193e8b9b2adbR655) ?

**Comment by chayanit** [comment]
> > Hello @laurenhay, why do we need new workflows for all 3 MC year? Is it supposed to be included in IB test only?
> 
> Hi @chayanit, do you mean [here](https://github.com/cms-sw/cmssw/pull/31795/files#diff-8c163203cbaa64f0e0e6aa2b16346cfab3eaf84a57a008c7fa37193e8b9b2adbR655) ?

Hi @rappoccio , I mean [https://github.com/cms-sw/cmssw/pull/31795/files#diff-09f871ad2ecabe8c1c5df4ce58cdfe28cb19dc1194361f8b29ebc7278869ca85R43-R44](here) and [https://github.com/cms-sw/cmssw/pull/31795/files#diff-b0dce787f15ea8526a7b5585e797ba4b9e17ba829550c8283b5b5f97cba58b04R3138-R3141](here). Why do we need to declare new workflows for 2016? Shouldn't it need only IB test which are included here instead? https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2017.py

**Comment by silviodonato** [comment]
> Hello @laurenhay, why do we need new workflows for all 3 MC year? Is it supposed to be included in IB test only?

@laurenhay

**Comment by laurenhay** [comment]
> > > Hello @laurenhay, why do we need new workflows for all 3 MC year? Is it supposed to be included in IB test only?
> > 
> > 
> > Hi @chayanit, do you mean [here](https://github.com/cms-sw/cmssw/pull/31795/files#diff-8c163203cbaa64f0e0e6aa2b16346cfab3eaf84a57a008c7fa37193e8b9b2adbR655) ?
> 
> Hi @rappoccio , I mean [https://github.com/cms-sw/cmssw/pull/31795/files#diff-09f871ad2ecabe8c1c5df4ce58cdfe28cb19dc1194361f8b29ebc7278869ca85R43-R44](here) and [https://github.com/cms-sw/cmssw/pull/31795/files#diff-b0dce787f15ea8526a7b5585e797ba4b9e17ba829550c8283b5b5f97cba58b04R3138-R3141](here). Why do we need to declare new workflows for 2016? Shouldn't it need only IB test which are included here instead? https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2017.py

I still don't know what you're referring to since we don't declare a workflow for 2016. The workflow we are upgrading for 2017 and 2018. The diff links you provided are broken as well.

File: Configuration/PyReleaseValidation/python/relval_pileup.py [hunk 1]
```diff
@@ -40,6 +40,8 @@
 workflows[25202.1]=['',['TTbar_13','DIGIUP15APVSimu_PU25','RECOUP15_PU25','HARVESTUP15_PU25']]
 workflows[25202.2]=['',['TTbar_13','DIGIUP15APVSimu_PU25','RECOUP15_PU25_HIPM','HARVESTUP15_PU25']]
 workflows[25202.15]=['',['TTbar_13','DIGIUP15_PU25','RECOUP15_PU25','HARVESTUP15_PU25','NANOUP15MC_PU25_JME']]
+workflows[25202.16]=['',['TTbar_13','DIGIUP15_PU25','RECOUP15_PU25','HARVESTUP15_PU25','NANOUP15MC_PU25_JMEAK8PF']]
+workflows[25202.17]=['',['TTbar_13','DIGIUP15_PU25','RECOUP15_PU25','HARVESTUP15_PU25','NANOUP15MC_PU25_JMEPF']]
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/relval_pileup.py:44]: what you declared here is for 2016 not 2017 and 2018 @laurenhay. If you don't want, you need to remove them

File: Configuration/PyReleaseValidation/python/relval_steps.py [hunk 2]
```diff
@@ -3135,6 +3135,10 @@ def gen2021HiMix(fragment,howMuch):
 steps['NANOUP15Had']=merge([{'--filein':'file:step4_inMINIAODSIM.root'},steps['NANOUP15']])
 steps['NANOUP15MC_PU25_JME']=merge([{'--customise':'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomNanoAOD_MC'},steps['NANOUP15']])
 steps['NANOUP15Data_PU25_JME']=merge([{'--customise':'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomNanoAOD_Data','--data':''},steps['NANOUP15']])
+steps['NANOUP15MC_PU25_JMEAK8PF']=merge([{'--customise':'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomAK8PFNanoAOD_MC'},steps['NANOUP15']])
+steps['NANOUP15Data_PU25_JMEAK8PF']=merge([{'--customise':'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomAK8PFNanoAOD_Data','--data':''},steps['NANOUP15']])
+steps['NANOUP15MC_PU25_JMEPF']=merge([{'--customise':'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomPFNanoAOD_MC'},steps['NANOUP15']])
+steps['NANOUP15Data_PU25_JMEPF']=merge([{'--customise':'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomPFNanoAOD_Data','--data':''},steps['NANOUP15']])
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/relval_steps.py:3141]: as well as these 4 lines here, they are for 2016 condition

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py [hunk 3]
```diff
@@ -646,6 +647,36 @@ def condition(self, fragment, stepList, key, hasHarvest):
     offset = 0.15,
 )
 
+class UpgradeWorkflow_JMEAK8PFNano(UpgradeWorkflow):
+    def setup_(self, step, stepName, stepDict, k, properties):
+        if 'Nano' in step:
+            stepDict[stepName][k] = merge([{'--customise': 'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomAK8PFNanoAOD_MC'}, stepDict[step][k]])
+    def condition(self, fragment, stepList, key, hasHarvest):
+        return fragment=="TTbar_13" and ('2017' in key or '2018' in key)
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:655]: This class is asking for 2017 and 2018 but why you need Run2? Shouldn't we target Run3?

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py [hunk 4]
```diff
@@ -646,6 +647,36 @@ def condition(self, fragment, stepList, key, hasHarvest):
...
+    suffix = '_JMEAK8PFNano',
+    offset = 0.16,
+)
+
+class UpgradeWorkflow_JMEPFNano(UpgradeWorkflow):
+    def setup_(self, step, stepName, stepDict, k, properties):
+        if 'Nano' in step:
+            stepDict[stepName][k] = merge([{'--customise': 'PhysicsTools/NanoAOD/custom_jme_cff.PrepJMECustomPFNanoAOD_MC'}, stepDict[step][k]])
+    def condition(self, fragment, stepList, key, hasHarvest):
+        return fragment=="TTbar_13" and ('2017' in key or '2018' in key)
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:670]: and same here

**Comment by rappoccio** [comment]
Thanks @chayanit. This PR is still under discussion of the scope so we will hold the 2016/2017/2018 workflows for now.

**Comment by silviodonato** [comment]
@laurenhay @rappoccio shall we close this PR?

**Comment by rappoccio** [comment]
Hi, Silvio,

That's fine.
Sal

**Comment by kpedro88** [comment]
please close

**Pull request closed without merging**